packages/@tailwindcss-upgrade/src/codemods/template/signatures.ts

```diff
@@ -0,0 +1,404 @@
+              node.nodes.unshift({ kind: 'selector', value: '*' })
+            }
+          })
+
+          if (changed) {
+            node.selector = SelectorParser.toCss(selectorAst)
+          }
+        }
+      })
+
+      // Compute the final signature, by generating the CSS for the variant
+      let signature = toCss(ast)
+      return signature
+    } catch {
+      // A unique symbol is returned to ensure that 2 signatures resulting in
+      // `null` are not considered equal.
+      return Symbol()
+    }
+  })
+})
+
+function temporarilyDisableThemeInline<T>(designSystem: DesignSystem, cb: () => T): T {
+  // Turn off `@theme inline` feature such that `@theme` and `@theme inline` are
+  // considered the same. The biggest motivation for this is referencing
+  // variables in another namespace that happen to contain the same value as the
+  // utility's own namespaces it is reading from.
+  //
+  // E.g.:
+  //
+  // The `max-w-*` utility doesn't read from the `--breakpoint-*` namespace.
+  // But it does read from the `--container-*` namespace. It also happens to
+  // be the case that `--breakpoint-md` and `--container-3xl` are the exact
+  // same value.
+  //
+  // If you then use the `max-w-(--breakpoint-md)` utility, inlining the
+  // variable would mean:
+  //  - `max-w-(--breakpoint-md)` → `max-width: 48rem;` → `max-w-3xl`
+  //  - `max-w-(--contianer-3xl)` → `max-width: 48rem;` → `max-w-3xl`
+  //
+  // Not inlining the variable would mean:
+  // - `max-w-(--breakpoint-md)` → `max-width: var(--breakpoint-md);` → `max-w-(--breakpoint-md)`
+  // - `max-w-(--container-3xl)` → `max-width: var(--container-3xl);` → `max-w-3xl`
+
+  // @ts-expect-error We are monkey-patching a method that's considered private
+  // in TypeScript
+  let originalGet = designSystem.theme.values.get
+
+  // Track all values with the inline option set, so we can restore them later.
+  let restorableInlineOptions = new Set<{ options: ThemeOptions }>()
+
+  // @ts-expect-error We are monkey-patching a method that's considered private
+  // in TypeScript
+  designSystem.theme.values.get = (key: string) => {
+    // @ts-expect-error We are monkey-patching a method that's considered private
+    // in TypeScript
+    let value = originalGet.call(designSystem.theme.values, key)
+    if (value === undefined) return value
+
+    // Remove `inline` if it was set
+    if (value.options & ThemeOptions.INLINE) {
+      restorableInlineOptions.add(value)
+      value.options &= ~ThemeOptions.INLINE
+    }
+
+    return value
+  }
+
+  // Run the callback with the `@theme inline` feature disabled
+  let result = cb()
+
+  // Restore the `@theme inline` to the original value
+  // @ts-expect-error We are monkey-patching a method that's private
+  designSystem.theme.values.get = originalGet
+
+  // Re-add the `inline` option, in case future lookups are done
+  for (let value of restorableInlineOptions) {
+    value.options |= ThemeOptions.INLINE
+  }
+
```

Comment:
should this be wrapped in a `try { … } finally { … }` for safety?

Reply:
Good idea!